tests/abi-utils/test_abi_utils.py

```diff
@@ -46,6 +70,14 @@ def test_fn_signature_to_4byte_selector(signature, expected):
     assert hex_selector == expected
 
 
+@pytest.mark.parametrize(
+    "abi,expected",
+    ((FN_ABI_D, "getOrderInfo((address,uint256,bytes,(address,uint256,bytes)))"),),
+)
+def test__abi_to_signature(abi, expected):
```

Comment:
Can you add a few more test cases? Some examples of things that might cause trouble:

1. When `'input'` key is missing entirely
2. A tuple that contains another tuple
3. An array of tuples

---

I'm a little nervous that it's going to get very hairy to support nested types. I wonder if we should just deprecate ABI methods in eth-utils and recommend using eth-abi instead.

Reply:
> Can you add a few more test cases? Some examples of things that might cause trouble:
> 
> 1. When `'input'` key is missing entirely
> 2. A tuple that contains another tuple
> 3. An array of tuples
> 
> I'm a little nervous that it's going to get very hairy to support nested types. I wonder if we should just deprecate ABI methods in eth-utils and recommend using eth-abi instead.

My original commit already had a test for a tuple that contains another tuple.  I've since added tests for missing `'input'` key and for an array of tuples.